pydis_site/apps/forms/discord.py

```diff
@@ -0,0 +1,143 @@
+
+
+__all__ = ("get_member", "get_roles")
+
+
+def fetch_and_update_roles() -> tuple[models.DiscordRole, ...]:
+    """Get information about roles from Discord."""
+    with httpx.Client() as client:
+        r = client.get(
+            f"{settings.DISCORD_API_BASE_URL}/guilds/{settings.DISCORD_GUILD_ID}/roles",
+            headers={"Authorization": f"Bot {settings.DISCORD_BOT_TOKEN}"},
+        )
+
+        r.raise_for_status()
+        return tuple(models.DiscordRole(**role) for role in r.json())
+
+
+def fetch_member_details(member_id: int) -> models.DiscordMember | None:
+    """Get a member by ID from the configured guild using the discord API."""
+    with httpx.Client() as client:
+        r = client.get(
+            f"{settings.DISCORD_API_BASE_URL}/guilds/{settings.DISCORD_GUILD_ID}/members/{member_id}",
+            headers={"Authorization": f"Bot {settings.DISCORD_BOT_TOKEN}"},
+        )
+
+        if r.status_code == 404:
+            return None
+
+        r.raise_for_status()
+        return models.DiscordMember(**r.json())
+
+
+def fetch_user_details(bearer_token: str) -> dict:
+    """Fetch information about the Discord user associated with the given ``bearer_token``."""
+    with httpx.Client() as client:
+        r = client.get(
+            f"{settings.DISCORD_API_BASE_URL}/users/@me",
+            headers={
+                "Authorization": f"Bearer {bearer_token}",
+            },
+        )
+
+        r.raise_for_status()
+
+        return r.json()
+
+
+def fetch_bearer_token(code: str, redirect: str, *, refresh: bool) -> dict:
+    """
+    Fetch an OAuth2 bearer token.
+
+    ## Arguments
+
+    - ``code``: The code or refresh token for the operation. Usually provided by Discord.
+    - ``redirect``: Where to redirect the client after successful login.
+
+    ## Keyword arguments
+
+    - ``refresh``: Whether to fetch a refresh token.
+    """
+    with httpx.Client() as client:
+        data = {
+            "client_id": settings.DISCORD_OAUTH2_CLIENT_ID,
+            "client_secret": settings.DISCORD_OAUTH2_CLIENT_SECRET,
+            "redirect_uri": f"{redirect}/callback",
+        }
+
+        if refresh:
+            data["grant_type"] = "refresh_token"
+            data["refresh_token"] = code
+        else:
+            data["grant_type"] = "authorization_code"
+            data["code"] = code
+
+        r = client.post(
+            f"{settings.DISCORD_API_BASE_URL}/oauth2/token",
+            headers={
+                "Content-Type": "application/x-www-form-urlencoded",
+            },
+            data=data,
+        )
+
+        r.raise_for_status()
+
+        return r.json()
+
+
+def get_roles(*, force_refresh: bool = False, stale_after: int = 60 * 60 * 24) -> tuple[models.DiscordRole, ...]:
+    """
+    Get a tuple of all roles from the cache, or discord API if not available.
+
+    ## Keyword arguments
+
+    - `force_refresh` (`bool`): Skip the cache and always update the roles from
```

Comment:
I think putting the type in the docs is redundant when we're using type annotations.